executor_test.go

```diff
@@ -3243,7 +3243,56 @@ func TestExecutor_Execute_Query_Error(t *testing.T) {
 			}
 		})
 	}
+}
+
+func TestExecutor_GroupByStrings(t *testing.T) {
+	c := test.MustRunCluster(t, 1)
+	defer c.Close()
+	c.CreateField(t, "istring", pilosa.IndexOptions{Keys: true}, "generals", pilosa.OptFieldKeys())
 
+	req := &pilosa.ImportRequest{
+		Index:      "istring",
+		Field:      "generals",
+		Shard:      0,
+		RowKeys:    []string{"r1", "r2", "r1", "r2", "r1", "r2", "r1", "r2", "r1", "r2"},
+		ColumnKeys: []string{"c1", "c2", "c3", "c4", "c5", "c6", "c7", "c8", "c9", "c10"},
+	}
+	if err := c[0].API.Import(context.Background(), req); err != nil {
+		t.Fatalf("importing: %v", err)
+	}
+
+	tests := []struct {
+		query    string
+		expected []pilosa.GroupCount
+	}{
+		{
+			query: "GroupBy(Rows(generals))",
+			expected: []pilosa.GroupCount{
+				{Group: []pilosa.FieldRow{{Field: "generals", RowID: 1, RowKey: "r1"}}, Count: 5},
+				{Group: []pilosa.FieldRow{{Field: "generals", RowID: 2, RowKey: "r2"}}, Count: 5},
+			},
+		},
+		{
+			query: "GroupBy(Rows(generals), filter=Row(generals=r2))",
+			expected: []pilosa.GroupCount{
+				{Group: []pilosa.FieldRow{{Field: "generals", RowID: 2, RowKey: "r2"}}, Count: 5},
+			},
+		},
+	}
+
+	for i, tst := range tests {
+		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
+			r, err := c[0].API.Query(context.Background(), &pilosa.QueryRequest{
+				Index: "istring",
+				Query: tst.query,
+			})
+			if err != nil {
+				t.Fatalf("got an error %v", err)
+			}
+			results := r.Results[0].([]pilosa.GroupCount)
+			test.CheckGroupBy(t, tst.expected, results)
+		})
+	}
 }
```

Comment:
I personally don't like table tests—they make all the error line numbers show up in the same line and they're just more difficult to debug in general. The actual test body isn't very large either so you could expand this into two separate subtests for probably the same number of lines.

Reply:
They are a bit easier when using the `t.Run` subtest functionality—*if* you give them a descriptive name. In most cases I name them with their index which works fine until you have more than 5 or so, and then it's a bit tough to figure out which one failed.

I like that they encourage adding new cases since it's pretty low effort, but I do think we could do a better job of naming each case to aid in debugging.

We should probably add some more verbiage in CONTRIBUTING around how to write tests, and how our tests are organized. We have enough table tests right now that I don't think we'd want to step away from that pattern, but 🤷‍♂ I'm open to it.